src/lpython/semantics/python_attribute_eval.h

```diff
@@ -74,13 +74,13 @@ struct AttributeHandler {
             throw SemanticError("Type name is not implemented yet.", loc);
         }
         std::string key = class_name + "@" + attr_name;
-        if (modify_attr_set.find(key) != modify_attr_set.end()) {
-            ASR::Variable_t* v = ASRUtils::EXPR2VAR(e);
-            if (v->m_intent == ASRUtils::intent_in) {
-                throw SemanticError("Modifying input function parameter `"
-                            + std::string(v->m_name) + "` is not allowed", loc);
-            }
-        }
+        // if (modify_attr_set.find(key) != modify_attr_set.end()) {
+        //     ASR::Variable_t* v = ASRUtils::EXPR2VAR(e);
+        //     if (v->m_intent == ASRUtils::intent_in) {
+        //         throw SemanticError("Modifying input function parameter `"
+        //                     + std::string(v->m_name) + "` is not allowed", loc);
+        //     }
+        // }
```

Comment:
@Shaikh-Ubaid could you please guide me on this? When handling `ListConstant`, we do not have any external symbol for it in the symbol table. How do we handle it?

Reply:
Could you share why you need an external symbol? Also, it would help if you could share what you are trying to compile?

Reply:
@Shaikh-Ubaid `ASRUtils::EXPR2VAR(e)` checks for `e` to be an external symbol above. How do we bypass or handle this for a `ListConstant` or for that matter any other related type?

Reply:
`ASRUtils::EXPR2VAR(e)` bypasses the external symbol node (if any) and gives you a reference to the variable node. After that it checks if the variable is of type `intent in`. In a function, for arguments with `intent in`, we do not allow updating them. To allow updating them, you need to explicitly specify using `Out`. For example:

```python
def my_func(l: Out[list[i32]]):
    pass
```

Reply:
Got it @Shaikh-Ubaid! We are trying to modify a `ListConstant` with this implementation, but this creation of a variable causes an assertion failure as `e` is not a `nullptr` here.

Reply:
Please revert this.

Reply:
> Please revert this.

@Shaikh-Ubaid reverting this will cause the exception as stated above. I think checking for `e` not being a `nullptr` should work. We had to handle that case I believe.

Reply:
Can you post an example of the exception that you are facing?

Reply:
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lpython$ ./src/bin/lpython ./examples/example.py
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  Binary file "/home/saurabh-kumar/Projects/System/lpython/src/bin/lpython", in _start()
  File "./csu/../csu/libc-start.c", line 360, in __libc_start_main_impl()
  File "./csu/../sysdeps/x86/libc-start.c", line 58, in __libc_start_call_main()
  File "/home/saurabh-kumar/Projects/System/lpython/src/bin/lpython.cpp", line 1999, in main()
    err = compile_python_using_llvm(arg_file, tmp_o, runtime_library_dir,
  File "/home/saurabh-kumar/Projects/System/lpython/src/bin/lpython.cpp", line 842, in compile_python_using_llvm()
    !(arg_c && compiler_options.po.disable_main), "__main__", infile);
  File "/home/saurabh-kumar/Projects/System/lpython/src/lpython/semantics/python_ast_to_asr.cpp", line 8212, in LCompilers::LPython::python_ast_to_asr(Allocator&, LCompilers::LocationManager&, LCompilers::SymbolTable*, LCompilers::LPython::AST::ast_t&, LCompilers::diag::Diagnostics&, LCompilers::CompilerOptions&, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool)
    ast_overload, allow_implicit_casting);
  File "/home/saurabh-kumar/Projects/System/lpython/src/lpython/semantics/python_ast_to_asr.cpp", line 8149, in LCompilers::LPython::body_visitor(Allocator&, LCompilers::LocationManager&, LCompilers::LPython::AST::Module_t const&, LCompilers::diag::Diagnostics&, LCompilers::ASR::asr_t*, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::map<int, LCompilers::ASR::symbol_t*, std::less<int>, std::allocator<std::pair<int const, LCompilers::ASR::symbol_t*> > >&, bool)
    b.visit_Module(ast);
  File "/home/saurabh-kumar/Projects/System/lpython/src/lpython/semantics/python_ast_to_asr.cpp", line 4796, in LCompilers::LPython::BodyVisitor::visit_Module(LCompilers::LPython::AST::Module_t const&)
    visit_stmt(*x.m_body[i]);
  File "/home/saurabh-kumar/Projects/System/lpython/src/lpython/python_ast.h", line 1883, in LCompilers::LPython::AST::BaseVisitor<LCompilers::LPython::BodyVisitor>::visit_stmt(LCompilers::LPython::AST::stmt_t const&)
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
  File "/home/saurabh-kumar/Projects/System/lpython/src/lpython/python_ast.h", line 1773, in visit_stmt_t<LCompilers::LPython::BodyVisitor>()
    case stmtType::Expr: { v.visit_Expr((const Expr_t &)x); return; }
  File "/home/saurabh-kumar/Projects/System/lpython/src/lpython/semantics/python_ast_to_asr.cpp", line 6583, in LCompilers::LPython::BodyVisitor::visit_Expr(LCompilers::LPython::AST::Expr_t const&)
    visit_Call(*c);
  File "/home/saurabh-kumar/Projects/System/lpython/src/lpython/semantics/python_ast_to_asr.cpp", line 7682, in LCompilers::LPython::BodyVisitor::visit_Call(LCompilers::LPython::AST::Call_t const&)
    visit_expr_list(x.m_args, x.n_args, args);
  File "/home/saurabh-kumar/Projects/System/lpython/src/lpython/semantics/python_ast_to_asr.cpp", line 613, in LCompilers::LPython::CommonVisitor<LCompilers::LPython::BodyVisitor>::visit_expr_list(LCompilers::LPython::AST::expr_t**, unsigned long, LCompilers::Vec<LCompilers::ASR::call_arg_t>&)
    this->visit_expr(*exprs[i]);
  File "/home/saurabh-kumar/Projects/System/lpython/src/lpython/python_ast.h", line 1910, in LCompilers::LPython::AST::BaseVisitor<LCompilers::LPython::BodyVisitor>::visit_expr(LCompilers::LPython::AST::expr_t const&)
    void visit_expr(const expr_t &b) { visit_expr_t(b, self()); }
  File "/home/saurabh-kumar/Projects/System/lpython/src/lpython/python_ast.h", line 1800, in visit_expr_t<LCompilers::LPython::BodyVisitor>()
    case exprType::Call: { v.visit_Call((const Call_t &)x); return; }
  File "/home/saurabh-kumar/Projects/System/lpython/src/lpython/semantics/python_ast_to_asr.cpp", line 7617, in LCompilers::LPython::BodyVisitor::visit_Call(LCompilers::LPython::AST::Call_t const&)
    handle_attribute(at, args, x.base.base.loc);
  File "/home/saurabh-kumar/Projects/System/lpython/src/lpython/semantics/python_ast_to_asr.cpp", line 7559, in LCompilers::LPython::BodyVisitor::handle_attribute(LCompilers::LPython::AST::Attribute_t*, LCompilers::Vec<LCompilers::ASR::call_arg_t>&, LCompilers::Location const&)
    handle_builtin_attribute(list_expr, at->m_attr, loc, eles);
  File "/home/saurabh-kumar/Projects/System/lpython/src/lpython/semantics/python_ast_to_asr.cpp", line 471, in LCompilers::LPython::CommonVisitor<LCompilers::LPython::BodyVisitor>::handle_builtin_attribute(LCompilers::ASR::expr_t*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::Location const&, LCompilers::Vec<LCompilers::ASR::expr_t*>&)
    tmp = attr_handler.get_attribute(s, attr_name, al, loc, args, diag);
  File "/home/saurabh-kumar/Projects/System/lpython/src/lpython/semantics/python_attribute_eval.h", line 78, in LCompilers::LPython::AttributeHandler::get_attribute(LCompilers::ASR::expr_t*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, Allocator&, LCompilers::Location const&, LCompilers::Vec<LCompilers::ASR::expr_t*>&, LCompilers::diag::Diagnostics&)
    ASR::Variable_t* v = ASRUtils::EXPR2VAR(e);
  File "/home/saurabh-kumar/Projects/System/lpython/src/libasr/asr_utils.h", line 275, in EXPR2VAR()
    ASR::down_cast<ASR::Var_t>(f)->m_v));
AssertFailed: is_a<T>(*f)
```

Reply:
Do you mind sharing the contents of your `./examples/example.py`?

Reply:
> Do you mind sharing the contents of your `./examples/example.py`?

@Shaikh-Ubaid it is the one I have shared in the PR description.

Reply:
The following should work. Please check.

```diff --git a/src/lpython/semantics/python_attribute_eval.h b/src/lpython/semantics/python_attribute_eval.h
index 941d5e4f3..aa0d37f42 100644
--- a/src/lpython/semantics/python_attribute_eval.h
+++ b/src/lpython/semantics/python_attribute_eval.h
@@ -74,13 +74,15 @@ struct AttributeHandler {
             throw SemanticError("Type name is not implemented yet.", loc);
         }
         std::string key = class_name + "@" + attr_name;
-        // if (modify_attr_set.find(key) != modify_attr_set.end()) {
-        //     ASR::Variable_t* v = ASRUtils::EXPR2VAR(e);
-        //     if (v->m_intent == ASRUtils::intent_in) {
-        //         throw SemanticError("Modifying input function parameter `"
-        //                     + std::string(v->m_name) + "` is not allowed", loc);
-        //     }
-        // }
+        if (modify_attr_set.find(key) != modify_attr_set.end()) {
+            if (ASR::is_a<ASR::Var_t>(*e)) {
+                ASR::Variable_t* v = ASRUtils::EXPR2VAR(e);
+                if (v->m_intent == ASRUtils::intent_in) {
+                    throw SemanticError("Modifying input function parameter `"
+                                + std::string(v->m_name) + "` is not allowed", loc);
+                }
+            }
+        }
         auto search = attribute_map.find(key);
         if (search != attribute_map.end()) {
             attribute_eval_callback cb = search->second;
```

Reply:
Thank you very much for this @Shaikh-Ubaid ! It compiles and works! I will make the changes.